### ISSUE TITLE
[[vk::push_constant]] variables are emitted as uniforms in SPIR-V

### ISSUE THREAD

**Issue by karroffel** [KhronosGroup/glslang#1629]
I am using HLSL for a project that uses vulkan and when trying to make use of push constants in a compute shader I noticed that the emitted SPIR-V code is incorrect.

Instead of annotating the push constant data as such it is marked as `Uniform`, which makes the creation of a compute pipeline fail.

The corresponding GLSL code emits the correct SPIR-V code.

<details>
<summary>HLSL code (not working) <a href="http://shader-playground.timjones.io/a973b824b76d533bfd61cfeeae3c78b2">playground</a></summary>

```hlsl
struct PushData {
    float val;
};

[[vk::push_constant]]
PushData pushData;

[[vk::binding(0)]]
RWStructuredBuffer<float> data;

void CSMain()
{
	data[0] = pushData.val;
}
```

</details>

<details>
<summary>GLSL code (working) <a href="http://shader-playground.timjones.io/4bff477558e5481d964127318bb4ba71">playground</a></summary>

```glsl
#version 450

layout(push_constant) uniform PushData {
    float val;
} push_data;

layout(binding = 0) buffer Output {
    float data[];
};

void main() {   
    data[0] = push_data.val;
}

```

</details>

**Comment by karroffel** [comment]
It seems like putting the struct into a `cbuffer` block does generate the correct `PushConstant` annotation.

From the [docs](https://github.com/Microsoft/DirectXShaderCompiler/blob/master/docs/SPIR-V.rst#push-constants) (and various tests) it seems like the provided code should work and a `cbuffer` should not be needed.

<details>
<summary>HLSL working with workaround <a href="http://shader-playground.timjones.io/8cde3a8541e931abb116d37eee04ec90">playground</a></summary>

```hlsl
struct PushData {
    float val;
};

[[vk::push_constant]]
cbuffer {
    PushData pushData;
};

[[vk::binding(0)]]
RWStructuredBuffer<float> data;

void CSMain()
{
	data[0] = pushData.val;
}
```

</details>

**Comment by johnkslang** [comment]
Right; the feature was designed in general (GLSL/HLSL/glslang-AST) as a refinement on uniform, rather than as a peer to uniform. But, I see your point that if push_constant is  requested, it could be made to "just work". For GLSL, that would require official language changes, but for HLSL, I suppose we could just do it.

**Comment by johnkslang** [comment]
If we add this feature, we'd want to be consistent with DXC as well.  Shader Playground reports that push_constant is ignored, but there is documentation for it.

**Comment by karroffel** [comment]
It seems like DXC does allow the attribute on any (one) global variable and also generates the proper SPIR-V code

[link to playground](http://shader-playground.timjones.io/4d3c1eee2061c9431fbe95ef299d396e)

There are also tests indicating that the attribute is made to work without a constant buffer wrapping it. (But it looks like wrapping in a constant buffer *also* works :smile: )

- [simple test](https://github.com/Microsoft/DirectXShaderCompiler/blob/285490211f2cbb44444a23d5f1c329881b96a43b/tools/clang/test/CodeGenSPIRV/vk.push-constant.hlsl)
- [test with explicit offsets](https://github.com/Microsoft/DirectXShaderCompiler/blob/285490211f2cbb44444a23d5f1c329881b96a43b/tools/clang/test/CodeGenSPIRV/vk.push-constant.offset.hlsl)

**Comment by Ryp** [comment]
This is the current way to get proper SPIR-V codegen on DXC and glslang:
```hlsl
#if defined(_DXC)
VK_PUSH_CONSTANT() CullPushConstants consts;
#else
VK_PUSH_CONSTANT() ConstantBuffer<CullPushConstants> consts;
#endif
```

DXC does not allow wrapping the struct in a `ConstantBuffer<>` and errors out:
```
error: 'push_constant' attribute only applies to global variables of struct type
```
I personally thing the DXC way makes more sense, but I'd be happy as long as the behavior becomes consistent between compilers :)

**Comment by tim-rex** [comment]
I found this issue largely by chance..  I come into this without in-depth prior knowledge of other shader languages, so I dn't have a great handle of what should be considered "correct" or not

Much like the issue already indicates.. I tried to make use of the following
```
[vk::push_constant] uniform float4x4 model;
```

and that resulted in a validation error
```
Vulkan Layer: VK_DEBUG_UTILS_MESSAGE_TYPE_VALIDATION_BIT_EXT : vkCreateGraphicsPipelines(): pCreateInfos[0].pStages[0] SPIR-V (VK_SHADER_STAGE_VERTEX_BIT) uses descriptor [Set 0, Binding 1, variable "globalParams"] (type VK_DESCRIPTOR_TYPE_UNIFORM_BUFFER or VK_DESCRIPTOR_TYPE_UNIFORM_BUFFER_DYNAMIC or VK_DESCRIPTOR_TYPE_INLINE_UNIFORM_BLOCK) but the VkDescriptorSetLayoutBinding::stageFlags was VK_SHADER_STAGE_FRAGMENT_BIT.
```

That implies that "globalParams" is an implicit buffer that I need to handle with a layout descriptor, and it would appear to be clashing with one I already have defined at that binding point..  

It seems like this is **not** intended behaviour, and that the syntax for push constants should be (at present) as per the workaround in this issue..
```
struct PushConstants
{
    float4x4 model;
};
[[vk::push_constant]] ConstantBuffer<PushConstants> push_constants;
```

That certainly appears to be working.. 


Does the [documentation](https://shader-slang.org/slang/user-guide/spirv-target-specific.html#push-constants) need to be updated, even if only to accomodate this workaround?
> You can make a global uniform parameter laid out as a push constant by using the [vk::push_constant] attribute on the uniform parameter.
